// web/src/components/Dashboard.tsx
import { useEffect, useState, Fragment } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Dialog, Transition } from '@headlessui/react';
import { Car, MapPin, RefreshCw } from 'lucide-react';
import { supabase } from '../lib/supabase';

export default function Dashboard() {
  const [user, setUser] = useState<any>(null);
  const [services, setServices] = useState<any[]>([]);
  const [loading, setLoading] = useState(true);
  const [errorMsg, setErrorMsg] = useState('');
  const navigate = useNavigate();

  // Modal para elegir vehículo
  const [showVehicleModal, setShowVehicleModal] = useState(false);
  const [selectedVehicle, setSelectedVehicle] = useState('MOTO');
  const [loadingBecomeDriver, setLoadingBecomeDriver] = useState(false);

  // ETAs por ID de servicio
  const [etas, setEtas] = useState<Record<string, any>>({});
  const [etaLoading, setEtaLoading] = useState<Record<string, boolean>>({});

  const loadServices = async () => {
    setLoading(true);
    setErrorMsg('');
    try {
      const token = localStorage.getItem('token');
      if (!token) {
        setLoading(false);
        navigate('/');
        return;
      }

      const userRes = await fetch(`${import.meta.env.VITE_BACKEND_URL}/users/me`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      const userData = await userRes.json();
      if (!userRes.ok) throw new Error(userData.error || 'Sesión inválida');
      setUser(userData.user);

      let servicesUrl = `${import.meta.env.VITE_BACKEND_URL}/services/my`;
      if (userData.user.role === 'DRIVER') {
        servicesUrl = `${import.meta.env.VITE_BACKEND_URL}/services/driver/my`;
      }

      const servicesRes = await fetch(servicesUrl, {
        headers: { Authorization: `Bearer ${token}` },
      });
      const servicesData = await servicesRes.json();
      if (!servicesRes.ok) throw new Error(servicesData.error || 'Error al cargar servicios');
      setServices(servicesData.services || []);

      setLoading(false);
    } catch (err: any) {
      console.error('Error cargando dashboard:', err);
      setErrorMsg(err.message || 'Error al cargar datos');
      localStorage.removeItem('token');
      navigate('/');
    }
  };

  useEffect(() => {
    loadServices();
  }, [navigate]);

  // Polling cada 10 segundos
  //useEffect(() => {
 //   const interval = setInterval(loadServices, 10000);
 //   return () => clearInterval(interval);
 // }, []);

// Notificaciones en tiempo real para conductor (alerta cuando llega oferta)
useEffect(() => {
  if (user?.role !== 'DRIVER') return;

  console.log('Iniciando suscripción Realtime para conductor ID:', user.id);
  // TODO: Mover esto al backend
  const channel = supabase
    .channel('services-changes')
    .on(
      'postgres_changes',
      {
        event: 'UPDATE',
        schema: 'public',
        table: 'service',
        filter: `driverId=eq.${user.id}`,
      },
      (payload: any) => {
        console.log('Evento Realtime recibido:', payload);
        if (payload.new.status === 'OFFERED') {
          alert('¡Nueva oferta de servicio recibida! Revisa tus asignados.');
          loadServices();
        }
      }
    )
    .subscribe((status: string) => {
      console.log('Estado de suscripción:', status);
      if (status === 'SUBSCRIBED') {
        console.log('¡Realtime suscrito correctamente!');
      } else if (status === 'CHANNEL_ERROR') {
        console.error('Error en canal Realtime');
      }
    });

  return () => {
    supabase.removeChannel(channel);
  };
}, [user?.role, user?.id, loadServices]);

  // Cargar ETAs
  useEffect(() => {
    if (!user || !services.length) return;

    services.forEach(s => {
      if (['OFFERED', 'ACCEPTED', 'ARRIVED'].includes(s.status) && s.driver && !etas[s.id]) {
        const fetchEta = async () => {
          setEtaLoading(prev => ({ ...prev, [s.id]: true }));
          try {
            const token = localStorage.getItem('token');
            const res = await fetch(`${import.meta.env.VITE_BACKEND_URL}/services/${s.id}/eta`, {
              headers: { Authorization: `Bearer ${token}` },
            });
            const data = await res.json();
            if (res.ok) {
              setEtas(prev => ({ ...prev, [s.id]: data }));
            }
          } catch (err) {
            console.error('Error ETA:', err);
          } finally {
            setEtaLoading(prev => ({ ...prev, [s.id]: false }));
          }
        };
        fetchEta();
      }
    });
  }, [services, user]);

  // Botones de acción para conductor
  const handleAcceptOffer = async (serviceId: string) => {
    try {
      const token = localStorage.getItem('token');
      const res = await fetch(`${import.meta.env.VITE_BACKEND_URL}/services/${serviceId}/accept`, {
        method: 'PATCH',
        headers: { Authorization: `Bearer ${token}` },
      });
      const data = await res.json();
      if (res.ok) {
        alert('Oferta aceptada');
        loadServices();
      } else {
        alert(data.error || 'Error al aceptar');
      }
    } catch (err) {
      alert('Error de conexión');
    }
  };

  const handleRejectOffer = async (serviceId: string) => {
    if (!window.confirm('¿Seguro que querés rechazar esta oferta?')) return;
    try {
      const token = localStorage.getItem('token');
      const res = await fetch(`${import.meta.env.VITE_BACKEND_URL}/services/${serviceId}/reject`, {
        method: 'PATCH',
        headers: { Authorization: `Bearer ${token}` },
      });
      const data = await res.json();
      if (res.ok) {
        alert('Oferta rechazada');
        loadServices();
      } else {
        alert(data.error || 'Error al rechazar');
      }
    } catch (err) {
      alert('Error de conexión');
    }
  };

  const handleMarkArrived = async (serviceId: string) => {
    if (!window.confirm('¿Confirmás que llegaste al punto de pickup?')) return;
    try {
      const token = localStorage.getItem('token');
      const res = await fetch(`${import.meta.env.VITE_BACKEND_URL}/services/${serviceId}/arrive`, {
        method: 'PATCH',
        headers: { Authorization: `Bearer ${token}` },
      });
      const data = await res.json();
      if (res.ok) {
        alert('Llegada marcada');
        loadServices();
      } else {
        alert(data.error || 'Error al marcar llegada');
      }
    } catch (err) {
      alert('Error de conexión');
    }
  };

  // NUEVA FUNCIÓN: Finalizar espera y calcular importe
  const handleFinishWait = async (serviceId: string) => {
    if (!window.confirm('¿Confirmás que finalizó la espera y querés calcular el importe?')) return;

    try {
      const token = localStorage.getItem('token');
      const res = await fetch(`${import.meta.env.VITE_BACKEND_URL}/services/${serviceId}/finish-wait`, {
        method: 'PATCH',
        headers: { Authorization: `Bearer ${token}` },
      });
      const data = await res.json();

      if (res.ok) {
        alert(`Espera finalizada. Importe calculado: ${data.service.importe} ARS`);
        loadServices(); // Refresca la lista
      } else {
        alert(data.error || 'Error al finalizar espera');
      }
    } catch (err) {
      alert('Error de conexión');
      console.error(err);
    }
  };

  // ¡Mueve ESTA función aquí arriba!
const handleBecomeDriver = async () => {
  setLoadingBecomeDriver(true);

  const token = localStorage.getItem('token');
  if (!token) {
    setErrorMsg('No estás logueado');
    setLoadingBecomeDriver(false);
    return;
  }

  try {
    const response = await fetch(`${import.meta.env.VITE_BACKEND_URL}/driver/profile`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ vehicleType: selectedVehicle }),
    });

    const data = await response.json();

    if (!response.ok) {
      setErrorMsg(data.error || 'Error al cambiar a conductor');
      setLoadingBecomeDriver(false);
      return;
    }

    setShowVehicleModal(false);
    alert(`¡Ahora sos conductor! Tipo: ${data.profile.vehicleType}`);
    window.location.reload(); // Refresca para actualizar rol
  } catch (err) {
    setErrorMsg('Error de conexión');
  } finally {
    setLoadingBecomeDriver(false);
  }
};

  if (loading) return <div className="flex h-screen items-center justify-center">
    <div className="animate-spin rounded-full h-16 w-16 border-t-4 border-white"></div>
  </div>;

  if (!user) return <div className="flex h-screen items-center justify-center text-center">
    <p className="text-2xl text-gray-400">{errorMsg || 'No estás logueado. '}<Link to="/" className="text-white underline">Iniciar sesión</Link></p>
  </div>;

  return (
    <div className="pt-16 pb-24">
      {/* Header */}
      <div className="mb-12">
        <h1 className="text-4xl font-bold mb-2">¡Hola, {user.email?.split('@')[0] || 'Usuario'}!</h1>
        <p className="text-gray-400 text-lg">Listo para moverte o ayudar a mover a otros?</p>
      </div>

      {/* Botón refrescar */}
      <div className="mb-6 flex justify-end">
        <button
          onClick={loadServices}
          disabled={loading}
          className="flex items-center gap-2 bg-gray-700 text-white px-4 py-2 rounded-xl hover:bg-gray-600 transition disabled:opacity-50"
        >
          <RefreshCw size={20} />
          Refrescar
        </button>
      </div>

      {/* Acciones principales */}
      <div className="grid grid-cols-1 gap-6 md:grid-cols-2 mb-12">
        {user.role === 'USER' && (
          <>
            <Link to="/request" className="bg-gray-800 rounded-2xl p-8 hover:bg-gray-700 transition">
              <div className="flex items-center gap-4 mb-4">
                <MapPin size={32} className="text-green-500" />
                <h2 className="text-2xl font-bold">Solicitar Servicio</h2>
              </div>
              <p className="text-gray-400">Un conductor llega en minutos</p>
            </Link>

            <button onClick={() => setShowVehicleModal(true)} className="bg-gray-800 rounded-2xl p-8 hover:bg-gray-700 transition">
              <div className="flex items-center gap-4 mb-4">
                <Car size={32} className="text-purple-500" />
                <h2 className="text-2xl font-bold">¿Querés ganar dinero?</h2>
              </div>
              <p className="text-gray-400">Conviértete en conductor</p>
            </button>
          </>
        )}
      </div>

      {/* Vista para conductores */}
      {user.role === 'DRIVER' && (
        <div className="mb-12">
          <h2 className="text-2xl font-bold mb-6">Mis servicios asignados</h2>
          {services.length === 0 ? (
            <p className="text-gray-400 text-center py-12">No tienes servicios asignados aún</p>
          ) : (
            <div className="space-y-6">
              {services.map((s: any) => (
                <div key={s.id} className="bg-gray-800 rounded-2xl p-6">
  <p className="font-bold text-lg mb-2">{s.type} solicitado por {s.requester?.email || 'Usuario'}</p>
  <p className="text-gray-400 text-sm mb-1">{new Date(s.requestedAt).toLocaleString()}</p>
  <p className="text-sm"><span className="font-medium">Estado:</span> {s.status}</p>
  <p className="text-sm"><span className="font-medium">Pickup:</span> Lat {s.pickupLat}, Lng {s.pickupLng}</p>

  {/* Mostrar importe si ya está completado */}
  {s.status === 'COMPLETED' && s.amount && (
    <p className="text-sm mt-2"><span className="font-medium">Importe calculado:</span> {s.amount} ARS</p>
  )}

  {/* Botones según estado */}
  {s.status === 'OFFERED' && (
    <div className="mt-4 flex gap-4">
      <button
        onClick={() => handleAcceptOffer(s.id)}
        className="flex-1 bg-green-600 text-white py-3 rounded-xl hover:bg-green-500 transition"
      >
        Aceptar oferta
      </button>
      <button
        onClick={() => handleRejectOffer(s.id)}
        className="flex-1 bg-red-600 text-white py-3 rounded-xl hover:bg-red-500 transition"
      >
        Rechazar
      </button>
    </div>
  )}

  {s.status === 'ACCEPTED' && (
    <button
      onClick={() => handleMarkArrived(s.id)}
      className="mt-4 w-full bg-blue-600 text-white py-3 rounded-xl hover:bg-blue-500 transition"
    >
      Marcar llegada
    </button>
  )}

  {s.status === 'ARRIVED' && (
    <button
      onClick={() => handleFinishWait(s.id)}
      className="mt-4 w-full bg-purple-600 text-white py-3 rounded-xl hover:bg-purple-500 transition"
    >
      Finalizar espera y calcular importe
    </button>
  )}
</div>
              ))}
            </div>
          )}
        </div>
      )}

      {/* Servicios recientes (solo para usuarios) */}
      {user.role !== 'DRIVER' && (
        <>
          <h2 className="text-2xl font-bold mb-6">Mis servicios recientes</h2>
          {services.length === 0 ? (
            <p className="text-gray-400 text-center py-12">No tienes servicios activos</p>
          ) : (
            <div className="space-y-6">
              {services.map((s: any) => (
                <div key={s.id} className="bg-gray-800 rounded-2xl p-6">
                  <p className="font-bold text-lg mb-2">{s.type}</p>
                  <p className="text-gray-400 text-sm mb-1">{new Date(s.requestedAt).toLocaleString()}</p>
                  <p className="text-sm"><span className="font-medium">Estado:</span> {s.status}</p>
                  {s.driver && <p className="text-sm"><span className="font-medium">Conductor:</span> {s.driver.email}</p>}
                  {s.amount && <p className="text-sm"><span className="font-medium">Importe:</span> {s.amount} ARS</p>}

                  {['OFFERED', 'ACCEPTED', 'ARRIVED'].includes(s.status) && s.driver && (
                    <div className="mt-4 p-3 bg-blue-900/30 rounded-xl">
                      {etaLoading[s.id] ? (
                        <p className="text-sm text-blue-400">Calculando ETA...</p>
                      ) : etas[s.id] ? (
                        <p className="text-sm font-medium">
                          {etas[s.id].message || 'Conductor asignado'}  
                          {etas[s.id].etaMinutes && ` - ETA: ${etas[s.id].etaMinutes} min`}
                          {etas[s.id].distanceKm && ` (${etas[s.id].distanceKm} km)`}
                        </p>
                      ) : (
                        <p className="text-sm text-gray-500">Sin ETA disponible (verifica ubicación del conductor)</p>
                      )}
                    </div>
                  )}
                </div>
              ))}
            </div>
          )}
        </>
      )}

      {/* Modal */}
      <Transition appear show={showVehicleModal} as={Fragment}>
        <Dialog as="div" className="relative z-50" onClose={() => setShowVehicleModal(false)}>
          <Transition.Child
            as={Fragment}
            enter="ease-out duration-300"
            enterFrom="opacity-0"
            enterTo="opacity-100"
            leave="ease-in duration-200"
            leaveFrom="opacity-100"
            leaveTo="opacity-0"
          >
            <div className="fixed inset-0 bg-black bg-opacity-75" />
          </Transition.Child>

          <div className="fixed inset-0 overflow-y-auto">
            <div className="flex min-h-full items-center justify-center p-4 text-center">
              <Transition.Child
                as={Fragment}
                enter="ease-out duration-300"
                enterFrom="opacity-0 scale-95"
                enterTo="opacity-100 scale-100"
                leave="ease-in duration-200"
                leaveFrom="opacity-100 scale-100"
                leaveTo="opacity-0 scale-95"
              >
                <Dialog.Panel className="w-full max-w-md transform overflow-hidden rounded-2xl bg-gray-900 p-8 text-left align-middle shadow-2xl transition-all border border-gray-700">
                  <Dialog.Title as="h3" className="text-2xl font-bold text-white mb-6 text-center">
                    Elige tu tipo de vehículo
                  </Dialog.Title>

                  <div className="grid grid-cols-1 gap-4 mb-8">
                    <button
                      onClick={() => setSelectedVehicle('MOTO')}
                      className={`flex items-center gap-4 p-5 rounded-xl transition-all ${
                        selectedVehicle === 'MOTO' 
                          ? 'bg-green-600 text-white ring-2 ring-green-400' 
                          : 'bg-gray-800 text-gray-300 hover:bg-gray-700'
                      }`}
                    >
                      <Car size={24} />
                      Moto
                    </button>

                    <button
                      onClick={() => setSelectedVehicle('TAXI')}
                      className={`flex items-center gap-4 p-5 rounded-xl transition-all ${
                        selectedVehicle === 'TAXI' 
                          ? 'bg-green-600 text-white ring-2 ring-green-400' 
                          : 'bg-gray-800 text-gray-300 hover:bg-gray-700'
                      }`}
                    >
                      <Car size={24} />
                      Taxi
                    </button>

                    <button
                      onClick={() => setSelectedVehicle('TRAFIC')}
                      className={`flex items-center gap-4 p-5 rounded-xl transition-all ${
                        selectedVehicle === 'TRAFIC' 
                          ? 'bg-green-600 text-white ring-2 ring-green-400' 
                          : 'bg-gray-800 text-gray-300 hover:bg-gray-700'
                      }`}
                    >
                      <Car size={24} />
                      Tráfico
                    </button>
                  </div>

                  <div className="flex gap-4">
                    <button
                      onClick={() => setShowVehicleModal(false)}
                      className="flex-1 bg-gray-700 text-gray-300 py-4 rounded-xl font-medium hover:bg-gray-600 transition"
                    >
                      Cancelar
                    </button>
                    <button
                      onClick={handleBecomeDriver}
                      disabled={loadingBecomeDriver}
                      className="flex-1 bg-green-500 text-black py-4 rounded-xl font-bold hover:bg-green-400 transition disabled:bg-gray-600 disabled:cursor-not-allowed"
                    >
                      {loadingBecomeDriver ? 'Procesando...' : 'Confirmar'}
                    </button>
                  </div>
                </Dialog.Panel>
              </Transition.Child>
            </div>
          </div>
        </Dialog>
      </Transition>
    </div>
  );
}
